File: python/app/app_peripherals_read_ft_sensor.py
#!/usr/bin/env python3
"""Mirror of app/app_peripherals_read_ft_sensor.cpp."""

from __future__ import annotations

import argparse
import time

from rcore_peripherals import FtSensor, FtSensorOptions


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Read F/T sensor samples")
    parser.add_argument("--serial-port", default="/dev/ttyUSB0")
    parser.add_argument("--sensor-type", default="xjc_serial")
    parser.add_argument("--baud-rate", type=int, default=460800)
    parser.add_argument("--sample-rate", type=float, default=1000.0)
    parser.add_argument("--read-buffer-size", type=int, default=512)
    parser.add_argument("--stale-timeout-ms", type=int, default=20)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    peripheral = FtSensor()
    options = FtSensorOptions(
        serial_port=args.serial_port,
        sensor_type=args.sensor_type,
        baud_rate=args.baud_rate,
        sample_rate_hz=args.sample_rate,
        read_buffer_size=args.read_buffer_size,
        stale_timeout_ms=args.stale_timeout_ms,
    )
    if not peripheral.Initialize(options):
        print("failed to initialize F/T peripheral", flush=True)
        return 1

    print(
        f"serial_port={options.serial_port} baud_rate={options.baud_rate} "
        f"type={options.sensor_type}",
        flush=True,
    )

    if not peripheral.Start():
        print("failed to start F/T peripheral", flush=True)
        return 1

    if not peripheral.WaitForFirstSample():
        info = peripheral.GetInfo()
        print(
            f"first valid frame timeout, frame_errors={info.frame_error_count}",
            flush=True,
        )
        return 1

    while True:
        sample = peripheral.GetSample()
        if sample is not None:
            print(
                sample.fx,
                sample.fy,
                sample.fz,
                sample.tx,
                sample.ty,
                sample.tz,
                flush=True,
            )
        time.sleep(0.001)


if __name__ == "__main__":
    raise SystemExit(main())

File: app/app_peripherals_read_ft_sensor.cpp
#include "peripherals/peripherals.hpp"

#include <chrono>
#include <cstdlib>
#include <iostream>
#include <string>
#include <thread>

namespace
{

bool ParseOptions(int argc, char **argv, smrcore::peripherals::FtSensorOptions &options)
{
    for (int i = 1; i < argc; ++i)
    {
        const std::string arg = argv[i];
        if (arg == "--serial-port" && i + 1 < argc)
        {
            options.serial_port = argv[++i];
        }
        else if (arg == "--sensor-type" && i + 1 < argc)
        {
            options.sensor_type = argv[++i];
        }
        else if (arg == "--baud-rate" && i + 1 < argc)
        {
            options.baud_rate = std::atoi(argv[++i]);
        }
        else if (arg == "--sample-rate" && i + 1 < argc)
        {
            options.sample_rate_hz = std::atof(argv[++i]);
        }
        else if (arg == "--stale-timeout-ms" && i + 1 < argc)
        {
            options.stale_timeout_ms = std::atoi(argv[++i]);
        }
        else if (arg == "--read-buffer-size" && i + 1 < argc)
        {
            options.read_buffer_size =
                static_cast<std::size_t>(std::strtoull(argv[++i], nullptr, 10));
        }
        else
        {
            std::cerr << "unknown argument: " << arg << '\n';
            return false;
        }
    }
    return true;
}

} // namespace

int main(int argc, char **argv)
{
    smrcore::peripherals::FtSensorOptions options;
    options.serial_port = "/dev/ttyUSB0";
    options.sample_rate_hz = 1000.0;
    if (!ParseOptions(argc, argv, options))
    {
        return 1;
    }

    smrcore::peripherals::FtSensor peripheral;
    if (!peripheral.Initialize(options))
    {
        std::cerr << "failed to initialize F/T peripheral\n";
        return 1;
    }

    std::cout << "serial_port=" << options.serial_port
              << " baud_rate=" << options.baud_rate << " type=" << options.sensor_type
              << '\n';

    if (!peripheral.Start())
    {
        std::cerr << "failed to start F/T peripheral\n";
        return 1;
    }

    if (!peripheral.WaitForFirstSample())
    {
        const auto info = peripheral.GetInfo();
        std::cerr << "first valid frame timeout, frame_errors="
                  << info.frame_error_count << '\n';
        return 1;
    }

    while (true)
    {
        if (auto sample = peripheral.GetSample(); sample.has_value())
        {
            std::cout << *sample << '\n';
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
}
